lib/posix-*: Add support for packet pipes & datagram unix sockets

andreittr opened the pull request:
### Description of changes

This changeset adds support for packet-mode pipes and datagram unix sockets, along with accompanying code to better make use of them. Specifically:
- Have `posix-fdio` honor the `O_DIRECT` flag set on open file decriptions and pass it to the I/O operations
- Replace implementation of `posix-pipe`
- Stream- or packet-mode is selected at runtime by the O_DIRECT flag passed to the `write` operation
- Add support for `AF_UNIX` sockets of type `SOCK_DGRAM` and `SOCK_SEQPACKET` (0-length datagrams are not supported)
- Add warning about not supporting 0-length datagrams

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

```
CONFIG_LIBPOSIX_PIPE=y
CONFIG_LIBPOSIX_PIPE_PACKET=y
CONFIG_LIBPOSIX_UNIXSOCKET=y
```

Pipe test snippet:
```c
int main(void)
{
	int p[2];
	char b[32];

	int r = pipe2(p, O_DIRECT|O_NONBLOCK);
	assert(!r);

	/* These writes are packets */
	printf("%zd\n", write(p[1], "qwert", 5));
	printf("%zd\n", write(p[1], "asdfgh", 6));
	printf("%zd\n", write(p[1], "12345678", 8));
	errno = 0;
	fcntl(p[1], F_SETFL, O_NONBLOCK);
	printf("-%d\n", errno);
	/* Switched to stream now */
	printf("%zd\n", write(p[1], "qwert", 5));
	printf("%zd\n", write(p[1], "asdfgh", 6));
	printf("%zd\n", write(p[1], "12345678", 8));
	errno = 0;
	fcntl(p[1], F_SETFL, O_DIRECT|O_NONBLOCK);
	printf("-%d\n", errno);
	/* Switched back to packet; however, stream still unread, writes append to it */
	printf("%zd\n", write(p[1], "qwert", 5));
	printf("%zd\n", write(p[1], "asdfgh", 6));
	printf("%zd\n", write(p[1], "12345678", 8));

	/* Read out */
	while ((r = read(p[0], b, 3)) > 0) {
		b[r] = '\0';
		puts(b);
	}
	printf("-%d\n", errno);

	/* Stream emptied, only now are writes actually packets */
	printf("%zd\n", write(p[1], "qwert", 5));
	printf("%zd\n", write(p[1], "asdfgh", 6));
	printf("%zd\n", write(p[1], "12345678", 8));
	printf("%zd\n", write(p[1], "qwert", 5));
	printf("%zd\n", write(p[1], "asdfgh", 6));
	printf("%zd\n", write(p[1], "12345678", 8));

	while ((r = read(p[0], b, 3)) > 0) {
		b[r] = '\0';
		puts(b);
	}
	printf("-%d\n", errno);

	return 0;
}
```

DGRAM socket test snippet (should print warning under Unikraft):
```c
int main(void)
{
	int sv[2];
	char b[32];

	int r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_NONBLOCK, 0, sv);
	assert(!r);

	printf("%zd\n", send(sv[1], "qwert", 5, 0));
	printf("%zd\n", send(sv[1], "asdfgh", 6, 0));
	printf("%zd\n", send(sv[1], "", 0, 0));
	printf("%zd\n", send(sv[1], "12345678", 8, 0));

	while ((r = read(sv[0], b, 3)) >= 0) {
		b[r] = '\0';
		puts(b);
	}
	printf("-%d\n", errno);

	return 0;
}
```

andreittr committed:
lib/posix-unixsocket: Add warning for 0-len dgrams

This change adds a warning to the send operation of unixsockets when a
datagram of zero length is attempted to be sent.

Currently unixsockets do not support 0-length datagrams and will
otherwise silently drop these packets. This is due to internal
implementation details that can be addressed when (and if) 0-length
unixsocket datagrams are relied on by workloads.
The warning then serves as a compatibility reminder in misbehaving apps.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

StefanJum reviewed (approved):
Works fine, thanks.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

andreistan26 reviewed (approved):
Looks good to me, thanks!

Reviewed-by: Andrei Stan [andreistan2003@gmail.com](mailto:andreistan2003@gmail.com)

andreittr commented:
Rebased on current staging.

andreittr commented:
This PR is on hold pending reconsideration of fcntl's handling of the O_DIRECT flag.

andreittr committed:
lib/posix-fdio: Pass O_DIRECT flag to I/O ops

This change makes posix-fdio pass the O_DIRECT flag, if set on the open
file mode, to the file read/write operations.

Checkpatch-Ignore: SPACING
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr committed:
lib/posix-pipe: Implement O_DIRECT (packet) pipes

This change replaces the implementation of pipe buffers with one that
supports both stream- and packet-mode communication, selected by using
the O_DIRECT flag.

Previously, this had been stubbed as an internal property of pipes, a
design that breaks the separation between files and open file
descriptions. This is now corrected, updating the API and its sole
consumer, posix-unixsocket.

Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: VOLATILE
Checkpatch-Ignore: TRAILING_STATEMENTS
Checkpatch-Ignore: SPACING
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
PR reopened; design revamped, as the previous one broke some clean abstractions:
- `fcntl` no longer passes any flags to drivers, these are _all_ handled at the open file description level (as they should)
- packet-mode is no longer a pipe file property, rather the open file description keeps track of the `O_DIRECT` flag
- pipe files honour the `O_DIRECT` flag on writes, selecting between stream- and packet-mode; this behaviour mirrors Linux
- `posix-pipe` API change
  - `posix-unixsocket` updated to use this new API

Reviews re-requested, since it's a reimplementation.

andreistan26 reviewed (approved):
Works great, thanks!

Reviewed-by: Andrei Stan [andreistan2003@gmail.com](mailto:andreistan2003@gmail.com)

StefanJum reviewed (approved):
All good, thanks.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
